gateway/src/apicast/policy/keycloak_scope_check/apicast-policy.json

```diff
@@ -0,0 +1,63 @@
+{
+  "$schema": "http://apicast.io/policy-v1/schema#manifest#",
+  "name": "Keycloak Scope Check",
```

Comment:
It is not really checking Scope, but Roles, right?

Reply:
@y-tabata this looks like one of the last blockers. This policy does not really do anything with OAuth Scopes as the name might imply.

Reply:
Since I think "role check" means only to check the particular roles are existing or not in the jwt and "scope check" means to check it has enough roles to access the particular resources or not, I believe "scope check" is better.

Reply:
But "scope" is a defined term in OIDC/OAuth terminology.  It is what Client is asking to access from the User. I'd try to not overload the term and give users false impressions this is related.

What about "Protect resources with Keycloak User and Realm Roles" ?

Reply:
According to [the document](https://www.keycloak.org/docs/4.0/server_admin/index.html#_role_scope_mappings), I understand that the "roles" in the access token is itself the "scope" which the user permitted to the client, so what we checks is the "scope".
How do you think?

> What about "Protect resources with Keycloak User and Realm Roles" ?

Your suggestion is not changing the policy name to "Keycloak Roles Check", but changing the summary like that?

Reply:
My suggestion would be to change the policy name to "Protect resources with Keycloak User and Realm Roles", but maybe that is too long to pass the validation.

The linked document explains how to limit "Scope" of the "Client" has to "User"'s "Roles".
It explains how a Client can have just limited Scope of access. It controls what roles will be given to the access token. 

This policy would still operate only on Roles.

Reply:
I agree, we should use "role" and not "scope".

The scope is the mean for the client (the consuming application) to ask to use the role of a user, or when dealing with the client credentials flow, the role of the client itself. 

But, there are also implicit roles (you do not have to set a scope to use it) and they are present in the token along with the explicit roles: https://www.keycloak.org/docs/3.3/server_admin/topics/roles/realm-roles.html

So, I would definitely prefer "role" and not "scope".

Reply:
I understand "role" is proper in Keycloak.
Thank you @mikz @nmasse-itix!